### ISSUE TITLE
Unable to use numbers as row records.

### ISSUE THREAD

**Issue by goldsam** [swimlane/ngx-datatable#741]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, post on Stackoverflow or Gitter
```

**Current behavior**
When using an array of integers for `rows`, a TypeError is thrown by the updateRows() method of src/components/body/body.component.ts when assigning an index to property `$$index` since numbers do not support properties.

**Expected behavior**
Number type row records should not be allowed and no exception should be thrown.  Furthermore, I would not expect any properties of my records to be altered in any way. 

**Reproduction of the problem**
Assign an array of numbers (e.g. [1,2,3] as the 'rows' data attribute value)

**What is the motivation / use case for changing the behavior?**
I want to lazily evaluate row column values as they scroll into view for rows which can be represented using an integral record identifier. My column cell templates would then lazily evaluate the displayed representation as a function (in the abstract sense of the word) of the record identifier.

I don't want to pay an unacceptably high computational and memory cost to generate object representations of each entry in my data set when the lightweight number representation should suffice.

**Please tell us about your environment:**
* **Table version:** 9.1.0
* **Angular version:** 4.03
* **Browser:** all
* **Language:** TypeScript 2.3

**Comment by adammedford** [comment]
I don't think this is a bug; it seems like you're attempting to hack together an unintended feature. Perhaps provide a Plunker demonstrating the constraints that necessitate this. What format do you receive your data in? The cost of generating object representations is rarely the bottleneck in my experience.

I'm not a part of this project; just trying to get you to flesh this out to a state that they will actually consider it in.

**Comment by goldsam** [comment]
My team is building a real-time performance-critical graphics application. We have invested a lot of effort implementing sophisticated high-performance indexing data structures. The API for these indices expose a typed array which indexes into an underlying object array.

At the rates which we visualize data (~30 Hz), building new object arrays to satisfy the ngx-datatable API is causing GC spikes resulting in devastating sputter. We have invested tremendous effort avoiding dynamic allocation of large objects. 

Ideally, we would like to be feed these indices directly to the table and have it yank data from our underlying data set. A better solution (any one consistent with using objects as rows) would be to support abstracting the data source API to an interface such as the following:

```typescript
interface DataProvider<T> {
  getRowCount(): number;
  getRow(index: number): T;
}
```

This is the approach taken by Slickgrid which we have used in the past with success. We would very much like to migrate to this library, as it would allow us to drop our jQuery dependency, it provides a cleaner API, and has thus far required a small fraction of the code required to integrate SlickGrid.

When I have time, I will open a new feature request to abstract the data source.
